src/mcpwm.rs

```diff
@@ -0,0 +1,527 @@
+    lowest_frequency: Hertz,
+    duty_mode: DutyMode,
+    counter_mode: CounterMode,
+    //deadtime: DeadtimeConfig,
+}
+
+impl OperatorConfig {
+    #[must_use]
+    pub fn frequency(mut self, frequency: Hertz) -> Self {
+        self.frequency = frequency;
+        self
+    }
+
+    #[cfg(idf_newer_than_or_equal_v4_4_0)]
+    #[must_use]
+    pub fn lowest_frequency(mut self, lowest_frequency: Hertz) -> Self {
+        self.lowest_frequency = lowest_frequency;
+        self
+    }
+
+    #[must_use]
+    pub fn duty_mode(mut self, duty_mode: DutyMode) -> Self {
+        self.duty_mode = duty_mode;
+        self
+    }
+
+    #[must_use]
+    pub fn counter_mode(mut self, counter_mode: CounterMode) -> Self {
+        self.counter_mode = counter_mode;
+        self
+    }
+
+    /*
+    #[must_use]
+    pub fn deadtime_config(mut self, deadtime_config: DeadtimeConfig) -> Self {
+        todo!()
+    }*/
+}
+
+impl Default for OperatorConfig {
+    fn default() -> Self {
+        Self {
+            frequency: 1000.Hz(),
+            duty_a: 50.0,
+            duty_b: 50.0,
+
+            #[cfg(idf_newer_than_or_equal_v4_4_0)]
+            lowest_frequency: 1.Hz(),
+
+            duty_mode: DutyMode::ActiveHigh,
+            counter_mode: CounterMode::Up,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct UnitZero;
+
+#[derive(Default)]
+pub struct UnitOne;
+
+pub type Duty = f32;
+const MAX_DUTY: Duty = 100.0;
+
+pub struct MCPWM<U: Unit> {
+    _unit: U,
+}
+
+impl<U: Unit> MCPWM<U> {
+    /// # Safety
+    ///
+    /// It is safe to instantiate this exactly one time per `Unit`.
+    unsafe fn new() -> Self {
+        Self {
+            _unit: U::default(),
+        }
+    }
+}
+
+pub trait Unit: Default {
+    fn unit() -> mcpwm_unit_t;
+}
+
+impl Unit for UnitZero {
+    fn unit() -> mcpwm_unit_t {
+        mcpwm_unit_t_MCPWM_UNIT_0
+    }
+}
+
+impl Unit for UnitOne {
+    fn unit() -> mcpwm_unit_t {
+        mcpwm_unit_t_MCPWM_UNIT_1
+    }
+}
+
+// TODO: How do we want fault module to fit into this?
+// TODO: How do we want capture module to fit into this?
+
+pub struct Mcpwm<U: Unit> {
+    _instance: MCPWM<U>,
+}
+
+impl<U: Unit> Mcpwm<U> {
+    pub fn new(instance: MCPWM<U>) -> Result<Self, EspError> {
+        Ok(Self {
+            _instance: instance,
+        })
+    }
+
+    // TODO: I do not at all understand the motivation behind exposing the group prescaler as
+    // "resolution"? From what i can see, duty is exposed as a percentage so we are not talking
+    // about the pulse length in terms of cycles as is the case for ledc, right? What am I missing?
+    // Anyway, my reasoning:
+    // * High res  => small prescaler  => better maintan accuracy at high frequency, unable to reach lower freq
+    // * Lower res => larger prescaler => worse accuracy at high frequency(still as good at low frequency),
+    //   able to reach lower freq
+    // Would it make more sense to expose it as "lowest reachable frequency"
```

Comment:
> Would it make more sense to expose it as "lowest reachable frequency"

I think this is the best way to express it, makes more sense this way imo.

> What happens if prescaler register value does not fit into the least significant 8 bits?

Probably bad things :D. We can either add the check here or PR to esp-idf and fix it there.